vsintegration/src/FSharp.Editor/FindReferencesService.fs

```diff
@@ -0,0 +1,121 @@
+// Copyright (c) Microsoft Corporation.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+namespace Microsoft.VisualStudio.FSharp.Editor
+
+open System
+open System.Collections.Immutable
+open System.Composition
+
+open Microsoft.CodeAnalysis
+open Microsoft.CodeAnalysis.Host.Mef
+open Microsoft.CodeAnalysis.Editor
+open Microsoft.CodeAnalysis.Editor.Host
+open Microsoft.CodeAnalysis.Navigation
+open Microsoft.CodeAnalysis.FindSymbols
+open Microsoft.CodeAnalysis.FindReferences
+
+open Microsoft.VisualStudio.FSharp.LanguageService
+
+open Microsoft.FSharp.Compiler.Range
+open Microsoft.FSharp.Compiler.SourceCodeServices
+
+[<ExportLanguageService(typeof<IStreamingFindReferencesService>, FSharpCommonConstants.FSharpLanguageName); Shared>]
+type internal FSharpFindReferencesService
+    [<ImportingConstructor>]
+    (
+        checkerProvider: FSharpCheckerProvider,
+        projectInfoManager: ProjectInfoManager
+    ) =
+
+    let findReferencedSymbolsAsync(document: Document, position: int, context: FindReferencesContext) : Async<unit> =
+        async {
+            let! sourceText = document.GetTextAsync(context.CancellationToken) |> Async.AwaitTask
+            let! textVersion = document.GetTextVersionAsync(context.CancellationToken) |> Async.AwaitTask
+            let checker = checkerProvider.Checker
+            let! options = projectInfoManager.TryGetOptionsForDocumentOrProject(document)
+            match options with
+            | Some options ->
+                let! _parseResults, checkResultsAnswer = checker.ParseAndCheckFileInProject(document.FilePath, hash textVersion, sourceText.ToString(), options)
+                let checkFileResults = 
+                    match checkResultsAnswer with
+                    | FSharpCheckFileAnswer.Aborted -> failwith "Compilation isn't complete yet"
```

Comment:
First, does this exception leak out from you to the caller (i.e. me)?

Second, does this exception represent something actually exceptional (i.e. some sort of bug that should never happen, and which you'd want to get a crash dump to investigate)?  Or is this something that can happen through normally?

If the former, then throwing is appropriate.  We'll send a crash dump to Watson, but we won't tear down VS.  If this happens normally though, then you don't want to throw an exception here.  We'll get millions of dumps for something that isn't really a bug.

So, if the former, it's ok to throw.  If the latter, you should not do this.